{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5cb4b669",
   "metadata": {},
   "source": [
    "# **`gos`** by example\n",
    "\n",
    "This notebook demonstrates the core features of `gos`:\n",
    "\n",
    "- Author declarative genomics visualizations which adhere to the [Gosling](http://gosling-lang.org/) JSON Specification \n",
    "- Render visualizations directly in notebook\n",
    "- Combine local, remote, and in-memory genomics data sources in visualizations\n",
    "- Control a Gosling visualization from Python\n",
    "\n",
    "Start by importing `gosling`. It is a convention to import as `gos` and then access the API through this namespace. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5df40d9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gosling as gos"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33e48339",
   "metadata": {},
   "source": [
    "## Creating `gos.Track`s and `gos.View`s\n",
    "\n",
    "**`gos`** exposes two fundemental building-blocks for genomics visualizatinos provided by the Gosling grammar:\n",
    "\n",
    "- `gos.Track`\n",
    "- `gos.View`\n",
    "\n",
    "A _Track_ is the core component of a genomics visualization that defines explict **transformations** and **mappings** of genomics data to **visual properties**. A _Track_ ay be composed with other _Tracks_ or **grouped** into a _View_ that share the same linked genomic domain.\n",
    "\n",
    "Every _Track_ therefore binds a data source. In `gos` we define a CSV dataset and bind it to a _Track_. We will start by loading a CSV containing UCSC hg38 cytoband information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f662f717",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Track({\n",
       "  data: {'type': 'csv', 'url': 'https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/UCSC.HG38.Human.CytoBandIdeogram.csv', 'chromosomeField': 'Chromosome', 'genomicFields': ['chromStart', 'chromEnd']},\n",
       "  height: 180,\n",
       "  mark: 'bar',\n",
       "  width: 800\n",
       "})"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "csv_url = \"https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/UCSC.HG38.Human.CytoBandIdeogram.csv\"\n",
    "\n",
    "data = gos.csv(\n",
    "    url=csv_url,\n",
    "    chromosomeField=\"Chromosome\",\n",
    "    genomicFields=[\"chromStart\", \"chromEnd\"],\n",
    ")\n",
    "\n",
    "gos.Track(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11344a81",
   "metadata": {},
   "source": [
    "The _Track_ above is bound to the genomics data, but we haven't declared how to map and transform these data to visual properties. For this, we will use the `gos.Track.mark_*()` and `gos.Track.encode()` method to declare which visual encoding to use and how to map and transform these data to this encoding. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "24f9fc5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Track({\n",
       "  color: Color({\n",
       "    domain: ['gneg', 'gpos25', 'gpos50', 'gpos75', 'gpos100', 'gvar'],\n",
       "    range: ['white', '#D9D9D9', '#979797', '#636363', 'black', '#A0A0F2'],\n",
       "    shorthand: 'Stain:N'\n",
       "  }),\n",
       "  data: {'type': 'csv', 'url': 'https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/UCSC.HG38.Human.CytoBandIdeogram.csv', 'chromosomeField': 'Chromosome', 'genomicFields': ['chromStart', 'chromEnd']},\n",
       "  height: 180,\n",
       "  mark: 'rect',\n",
       "  size: SizeValue({\n",
       "    value: 20\n",
       "  }),\n",
       "  stroke: StrokeValue({\n",
       "    value: 'gray'\n",
       "  }),\n",
       "  strokeWidth: StrokeWidthValue({\n",
       "    value: 0.5\n",
       "  }),\n",
       "  width: 800,\n",
       "  x: X({\n",
       "    axis: 'top',\n",
       "    shorthand: 'chromStart:G'\n",
       "  }),\n",
       "  xe: Xe({\n",
       "    shorthand: 'chromEnd:G'\n",
       "  })\n",
       "})"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "track = gos.Track(data).mark_rect().encode(\n",
    "    # defines start and end of rectangle mark\n",
    "    x=gos.X(\"chromStart:G\", axis=\"top\"),\n",
    "    xe=gos.Xe(\"chromEnd:G\"),\n",
    "    # defines how to map values in \"Stain\" column to colors\n",
    "    color=gos.Color(\n",
    "        \"Stain:N\", \n",
    "        domain=[\"gneg\", \"gpos25\", \"gpos50\", \"gpos75\", \"gpos100\", \"gvar\"],\n",
    "        range=[\"white\", \"#D9D9D9\", \"#979797\", \"#636363\", \"black\", \"#A0A0F2\"]\n",
    "    ),\n",
    "    # customize the style of the visual marks. \n",
    "    size=gos.value(20),\n",
    "    stroke=gos.value(\"gray\"),\n",
    "    strokeWidth=gos.value(0.5)\n",
    ")\n",
    "track"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87d21d66",
   "metadata": {},
   "source": [
    "Our `gos.Track` now is fully specified, however, the Gosling grammar defines the root of every visualization as a _View_, which may contain one or more _Tracks_. Therefore in order to complete a Gosling specification for the track in isolation, we use the `gos.Track.view()` method to cast the track within a `gos.View`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "3c67f1c4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<!DOCTYPE html>\n",
       "<html>\n",
       "<head>\n",
       "  <style>.error { color: red; }</style>\n",
       "  <link rel=\"stylesheet\" href=\"https://unpkg.com/higlass@1.11/dist/hglib.css\">\n",
       "</head>\n",
       "<body>\n",
       "  <div id=\"jupyter-gosling-ed6752e20cd84b119408218ffa99a305\"></div>\n",
       "  <script type=\"module\">\n",
       "\n",
       "    async function loadScript(src) {\n",
       "        return new Promise(resolve => {\n",
       "            const script = document.createElement('script');\n",
       "            script.onload = resolve;\n",
       "            script.src = src;\n",
       "            script.async = false;\n",
       "            document.head.appendChild(script);\n",
       "        });\n",
       "    }\n",
       "\n",
       "    async function loadGosling() {\n",
       "        // Manually load scripts from window namespace since requirejs might not be\n",
       "        // available in all browser environments.\n",
       "\n",
       "        // https://github.com/DanielHreben/requirejs-toggle\n",
       "        if (!window.gosling) {\n",
       "\n",
       "            // https://github.com/DanielHreben/requirejs-toggle\n",
       "            window.__requirejsToggleBackup = {\n",
       "                define: window.define,\n",
       "                require: window.require,\n",
       "                requirejs: window.requirejs,\n",
       "            };\n",
       "\n",
       "            for (const field of Object.keys(window.__requirejsToggleBackup)) {\n",
       "                window[field] = undefined;\n",
       "            }\n",
       "\n",
       "            // load dependencies sequentially\n",
       "            const sources = [\n",
       "                \"https://unpkg.com/react@17/umd/react.production.min.js\",\n",
       "                \"https://unpkg.com/react-dom@17/umd/react-dom.production.min.js\",\n",
       "                \"https://unpkg.com/pixi.js@6/dist/browser/pixi.min.js\",\n",
       "                \"https://unpkg.com/higlass@1.11/dist/hglib.js\",\n",
       "                \"https://unpkg.com/gosling.js@0.9.18/dist/gosling.js\",\n",
       "            ];\n",
       "\n",
       "            for (const src of sources) await loadScript(src);\n",
       "\n",
       "            // restore requirejs after scripts have loaded\n",
       "            Object.assign(window, window.__requirejsToggleBackup);\n",
       "            delete window.__requirejsToggleBackup;\n",
       "\n",
       "        }\n",
       "\n",
       "        return window.gosling;\n",
       "    };\n",
       "\n",
       "    var el = document.getElementById('jupyter-gosling-ed6752e20cd84b119408218ffa99a305');\n",
       "    var spec = {\"tracks\": [{\"data\": {\"type\": \"csv\", \"url\": \"https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/UCSC.HG38.Human.CytoBandIdeogram.csv\", \"chromosomeField\": \"Chromosome\", \"genomicFields\": [\"chromStart\", \"chromEnd\"]}, \"height\": 180, \"mark\": \"rect\", \"width\": 800, \"color\": {\"domain\": [\"gneg\", \"gpos25\", \"gpos50\", \"gpos75\", \"gpos100\", \"gvar\"], \"field\": \"Stain\", \"range\": [\"white\", \"#D9D9D9\", \"#979797\", \"#636363\", \"black\", \"#A0A0F2\"], \"type\": \"nominal\"}, \"size\": {\"value\": 20}, \"stroke\": {\"value\": \"gray\"}, \"strokeWidth\": {\"value\": 0.5}, \"x\": {\"axis\": \"top\", \"field\": \"chromStart\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"chromEnd\", \"type\": \"genomic\"}}]};\n",
       "    var opt = {\"padding\": 0, \"theme\": null};\n",
       "\n",
       "    loadGosling()\n",
       "        .then(gosling => gosling.embed(el, spec, opt))\n",
       "        .catch(err => {\n",
       "            el.innerHTML = `<div class=\"error\">\n",
       "    <p>JavaScript Error: ${error.message}</p>\n",
       "    <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>\n",
       "</div>`;\n",
       "            throw error;\n",
       "        });\n",
       "  </script>\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "View({\n",
       "  tracks: [Track({\n",
       "    color: Color({\n",
       "      domain: ['gneg', 'gpos25', 'gpos50', 'gpos75', 'gpos100', 'gvar'],\n",
       "      field: 'Stain',\n",
       "      range: ['white', '#D9D9D9', '#979797', '#636363', 'black', '#A0A0F2'],\n",
       "      type: 'nominal'\n",
       "    }),\n",
       "    data: {'type': 'csv', 'url': 'https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/UCSC.HG38.Human.CytoBandIdeogram.csv', 'chromosomeField': 'Chromosome', 'genomicFields': ['chromStart', 'chromEnd']},\n",
       "    height: 180,\n",
       "    mark: 'rect',\n",
       "    size: SizeValue({\n",
       "      value: 20\n",
       "    }),\n",
       "    stroke: StrokeValue({\n",
       "      value: 'gray'\n",
       "    }),\n",
       "    strokeWidth: StrokeWidthValue({\n",
       "      value: 0.5\n",
       "    }),\n",
       "    width: 800,\n",
       "    x: X({\n",
       "      axis: 'top',\n",
       "      field: 'chromStart',\n",
       "      type: 'genomic'\n",
       "    }),\n",
       "    xe: Xe({\n",
       "      field: 'chromEnd',\n",
       "      type: 'genomic'\n",
       "    })\n",
       "  })]\n",
       "})"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "track.view()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9e4ab3c",
   "metadata": {},
   "source": [
    "Additional parameters for the `gos.View` can be passed in as well for convenience. We can easily set a `title` and `xDomain` for our _View_, initializing the visualization to display \"chr1\". \n",
    "\n",
    "> Notice how we reuse the `track` instance to create new, modified views. This is a common pattern in **`gos`**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "5ece39d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<!DOCTYPE html>\n",
       "<html>\n",
       "<head>\n",
       "  <style>.error { color: red; }</style>\n",
       "  <link rel=\"stylesheet\" href=\"https://unpkg.com/higlass@1.11/dist/hglib.css\">\n",
       "</head>\n",
       "<body>\n",
       "  <div id=\"jupyter-gosling-1579f6e751de43b7989c03d2ecb2b867\"></div>\n",
       "  <script type=\"module\">\n",
       "\n",
       "    async function loadScript(src) {\n",
       "        return new Promise(resolve => {\n",
       "            const script = document.createElement('script');\n",
       "            script.onload = resolve;\n",
       "            script.src = src;\n",
       "            script.async = false;\n",
       "            document.head.appendChild(script);\n",
       "        });\n",
       "    }\n",
       "\n",
       "    async function loadGosling() {\n",
       "        // Manually load scripts from window namespace since requirejs might not be\n",
       "        // available in all browser environments.\n",
       "\n",
       "        // https://github.com/DanielHreben/requirejs-toggle\n",
       "        if (!window.gosling) {\n",
       "\n",
       "            // https://github.com/DanielHreben/requirejs-toggle\n",
       "            window.__requirejsToggleBackup = {\n",
       "                define: window.define,\n",
       "                require: window.require,\n",
       "                requirejs: window.requirejs,\n",
       "            };\n",
       "\n",
       "            for (const field of Object.keys(window.__requirejsToggleBackup)) {\n",
       "                window[field] = undefined;\n",
       "            }\n",
       "\n",
       "            // load dependencies sequentially\n",
       "            const sources = [\n",
       "                \"https://unpkg.com/react@17/umd/react.production.min.js\",\n",
       "                \"https://unpkg.com/react-dom@17/umd/react-dom.production.min.js\",\n",
       "                \"https://unpkg.com/pixi.js@6/dist/browser/pixi.min.js\",\n",
       "                \"https://unpkg.com/higlass@1.11/dist/hglib.js\",\n",
       "                \"https://unpkg.com/gosling.js@0.9.18/dist/gosling.js\",\n",
       "            ];\n",
       "\n",
       "            for (const src of sources) await loadScript(src);\n",
       "\n",
       "            // restore requirejs after scripts have loaded\n",
       "            Object.assign(window, window.__requirejsToggleBackup);\n",
       "            delete window.__requirejsToggleBackup;\n",
       "\n",
       "        }\n",
       "\n",
       "        return window.gosling;\n",
       "    };\n",
       "\n",
       "    var el = document.getElementById('jupyter-gosling-1579f6e751de43b7989c03d2ecb2b867');\n",
       "    var spec = {\"tracks\": [{\"data\": {\"type\": \"csv\", \"url\": \"https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/UCSC.HG38.Human.CytoBandIdeogram.csv\", \"chromosomeField\": \"Chromosome\", \"genomicFields\": [\"chromStart\", \"chromEnd\"]}, \"height\": 180, \"mark\": \"rect\", \"width\": 800, \"color\": {\"domain\": [\"gneg\", \"gpos25\", \"gpos50\", \"gpos75\", \"gpos100\", \"gvar\"], \"field\": \"Stain\", \"range\": [\"white\", \"#D9D9D9\", \"#979797\", \"#636363\", \"black\", \"#A0A0F2\"], \"type\": \"nominal\"}, \"size\": {\"value\": 20}, \"stroke\": {\"value\": \"gray\"}, \"strokeWidth\": {\"value\": 0.5}, \"x\": {\"axis\": \"top\", \"field\": \"chromStart\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"chromEnd\", \"type\": \"genomic\"}}], \"title\": \"Gos is awesome!\", \"xDomain\": {\"chromosome\": \"chr1\"}};\n",
       "    var opt = {\"padding\": 0, \"theme\": null};\n",
       "\n",
       "    loadGosling()\n",
       "        .then(gosling => gosling.embed(el, spec, opt))\n",
       "        .catch(err => {\n",
       "            el.innerHTML = `<div class=\"error\">\n",
       "    <p>JavaScript Error: ${error.message}</p>\n",
       "    <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>\n",
       "</div>`;\n",
       "            throw error;\n",
       "        });\n",
       "  </script>\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "View({\n",
       "  title: 'Gos is awesome!',\n",
       "  tracks: [Track({\n",
       "    color: Color({\n",
       "      domain: ['gneg', 'gpos25', 'gpos50', 'gpos75', 'gpos100', 'gvar'],\n",
       "      field: 'Stain',\n",
       "      range: ['white', '#D9D9D9', '#979797', '#636363', 'black', '#A0A0F2'],\n",
       "      type: 'nominal'\n",
       "    }),\n",
       "    data: {'type': 'csv', 'url': 'https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/UCSC.HG38.Human.CytoBandIdeogram.csv', 'chromosomeField': 'Chromosome', 'genomicFields': ['chromStart', 'chromEnd']},\n",
       "    height: 180,\n",
       "    mark: 'rect',\n",
       "    size: SizeValue({\n",
       "      value: 20\n",
       "    }),\n",
       "    stroke: StrokeValue({\n",
       "      value: 'gray'\n",
       "    }),\n",
       "    strokeWidth: StrokeWidthValue({\n",
       "      value: 0.5\n",
       "    }),\n",
       "    width: 800,\n",
       "    x: X({\n",
       "      axis: 'top',\n",
       "      field: 'chromStart',\n",
       "      type: 'genomic'\n",
       "    }),\n",
       "    xe: Xe({\n",
       "      field: 'chromEnd',\n",
       "      type: 'genomic'\n",
       "    })\n",
       "  })],\n",
       "  xDomain: GenomicDomain({\n",
       "    chromosome: 'chr1'\n",
       "  })\n",
       "})"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "track.view(\n",
    "    title=\"Gos is awesome!\",\n",
    "    xDomain=gos.GenomicDomain(chromosome=\"chr1\"),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2cbf8879",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<!DOCTYPE html>\n",
       "<html>\n",
       "<head>\n",
       "  <style>.error { color: red; }</style>\n",
       "  <link rel=\"stylesheet\" href=\"https://unpkg.com/higlass@1.11/dist/hglib.css\">\n",
       "</head>\n",
       "<body>\n",
       "  <div id=\"jupyter-gosling-26de1716216443fda48def9c33d72111\"></div>\n",
       "  <script type=\"module\">\n",
       "\n",
       "    async function loadScript(src) {\n",
       "        return new Promise(resolve => {\n",
       "            const script = document.createElement('script');\n",
       "            script.onload = resolve;\n",
       "            script.src = src;\n",
       "            script.async = false;\n",
       "            document.head.appendChild(script);\n",
       "        });\n",
       "    }\n",
       "\n",
       "    async function loadGosling() {\n",
       "        // Manually load scripts from window namespace since requirejs might not be\n",
       "        // available in all browser environments.\n",
       "\n",
       "        // https://github.com/DanielHreben/requirejs-toggle\n",
       "        if (!window.gosling) {\n",
       "\n",
       "            // https://github.com/DanielHreben/requirejs-toggle\n",
       "            window.__requirejsToggleBackup = {\n",
       "                define: window.define,\n",
       "                require: window.require,\n",
       "                requirejs: window.requirejs,\n",
       "            };\n",
       "\n",
       "            for (const field of Object.keys(window.__requirejsToggleBackup)) {\n",
       "                window[field] = undefined;\n",
       "            }\n",
       "\n",
       "            // load dependencies sequentially\n",
       "            const sources = [\n",
       "                \"https://unpkg.com/react@17/umd/react.production.min.js\",\n",
       "                \"https://unpkg.com/react-dom@17/umd/react-dom.production.min.js\",\n",
       "                \"https://unpkg.com/pixi.js@6/dist/browser/pixi.min.js\",\n",
       "                \"https://unpkg.com/higlass@1.11/dist/hglib.js\",\n",
       "                \"https://unpkg.com/gosling.js@0.9.18/dist/gosling.js\",\n",
       "            ];\n",
       "\n",
       "            for (const src of sources) await loadScript(src);\n",
       "\n",
       "            // restore requirejs after scripts have loaded\n",
       "            Object.assign(window, window.__requirejsToggleBackup);\n",
       "            delete window.__requirejsToggleBackup;\n",
       "\n",
       "        }\n",
       "\n",
       "        return window.gosling;\n",
       "    };\n",
       "\n",
       "    var el = document.getElementById('jupyter-gosling-26de1716216443fda48def9c33d72111');\n",
       "    var spec = {\"alignment\": \"stack\", \"tracks\": [{\"alignment\": \"overlay\", \"tracks\": [{\"data\": {\"type\": \"beddb\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation\", \"genomicFields\": [{\"index\": 1, \"name\": \"start\"}, {\"index\": 2, \"name\": \"end\"}], \"valueFields\": [{\"index\": 5, \"name\": \"strand\", \"type\": \"nominal\"}, {\"index\": 3, \"name\": \"name\", \"type\": \"nominal\"}], \"exonIntervalFields\": [{\"index\": 12, \"name\": \"start\"}, {\"index\": 13, \"name\": \"end\"}]}, \"mark\": \"text\", \"dataTransform\": [{\"type\": \"filter\", \"field\": \"type\", \"oneOf\": [\"gene\"]}], \"style\": {\"dy\": -15}, \"text\": {\"field\": \"name\", \"type\": \"nominal\"}, \"x\": {\"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}}, {\"data\": {\"type\": \"beddb\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation\", \"genomicFields\": [{\"index\": 1, \"name\": \"start\"}, {\"index\": 2, \"name\": \"end\"}], \"valueFields\": [{\"index\": 5, \"name\": \"strand\", \"type\": \"nominal\"}, {\"index\": 3, \"name\": \"name\", \"type\": \"nominal\"}], \"exonIntervalFields\": [{\"index\": 12, \"name\": \"start\"}, {\"index\": 13, \"name\": \"end\"}]}, \"mark\": \"triangleRight\", \"dataTransform\": [{\"type\": \"filter\", \"field\": \"type\", \"oneOf\": [\"gene\"]}, {\"type\": \"filter\", \"field\": \"strand\", \"oneOf\": [\"+\"]}], \"size\": {\"value\": 15}, \"x\": {\"axis\": \"top\", \"field\": \"end\", \"type\": \"genomic\"}}, {\"data\": {\"type\": \"beddb\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation\", \"genomicFields\": [{\"index\": 1, \"name\": \"start\"}, {\"index\": 2, \"name\": \"end\"}], \"valueFields\": [{\"index\": 5, \"name\": \"strand\", \"type\": \"nominal\"}, {\"index\": 3, \"name\": \"name\", \"type\": \"nominal\"}], \"exonIntervalFields\": [{\"index\": 12, \"name\": \"start\"}, {\"index\": 13, \"name\": \"end\"}]}, \"mark\": \"triangleLeft\", \"dataTransform\": [{\"type\": \"filter\", \"field\": \"type\", \"oneOf\": [\"gene\"]}, {\"type\": \"filter\", \"field\": \"strand\", \"oneOf\": [\"-\"]}], \"size\": {\"value\": 15}, \"style\": {\"align\": \"right\"}, \"x\": {\"field\": \"start\", \"type\": \"genomic\"}}, {\"data\": {\"type\": \"beddb\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation\", \"genomicFields\": [{\"index\": 1, \"name\": \"start\"}, {\"index\": 2, \"name\": \"end\"}], \"valueFields\": [{\"index\": 5, \"name\": \"strand\", \"type\": \"nominal\"}, {\"index\": 3, \"name\": \"name\", \"type\": \"nominal\"}], \"exonIntervalFields\": [{\"index\": 12, \"name\": \"start\"}, {\"index\": 13, \"name\": \"end\"}]}, \"mark\": \"rect\", \"dataTransform\": [{\"type\": \"filter\", \"field\": \"type\", \"oneOf\": [\"exon\"]}], \"size\": {\"value\": 15}, \"x\": {\"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}}, {\"data\": {\"type\": \"beddb\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation\", \"genomicFields\": [{\"index\": 1, \"name\": \"start\"}, {\"index\": 2, \"name\": \"end\"}], \"valueFields\": [{\"index\": 5, \"name\": \"strand\", \"type\": \"nominal\"}, {\"index\": 3, \"name\": \"name\", \"type\": \"nominal\"}], \"exonIntervalFields\": [{\"index\": 12, \"name\": \"start\"}, {\"index\": 13, \"name\": \"end\"}]}, \"mark\": \"rule\", \"dataTransform\": [{\"type\": \"filter\", \"field\": \"type\", \"oneOf\": [\"gene\"]}, {\"type\": \"filter\", \"field\": \"strand\", \"oneOf\": [\"+\"]}], \"strokeWidth\": {\"value\": 3}, \"style\": {\"linePattern\": {\"type\": \"triangleRight\", \"size\": 5}}, \"x\": {\"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}}, {\"data\": {\"type\": \"beddb\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation\", \"genomicFields\": [{\"index\": 1, \"name\": \"start\"}, {\"index\": 2, \"name\": \"end\"}], \"valueFields\": [{\"index\": 5, \"name\": \"strand\", \"type\": \"nominal\"}, {\"index\": 3, \"name\": \"name\", \"type\": \"nominal\"}], \"exonIntervalFields\": [{\"index\": 12, \"name\": \"start\"}, {\"index\": 13, \"name\": \"end\"}]}, \"mark\": \"rule\", \"dataTransform\": [{\"type\": \"filter\", \"field\": \"type\", \"oneOf\": [\"gene\"]}, {\"type\": \"filter\", \"field\": \"strand\", \"oneOf\": [\"-\"]}], \"strokeWidth\": {\"value\": 3}, \"style\": {\"linePattern\": {\"type\": \"triangleLeft\", \"size\": 5}}, \"x\": {\"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}}], \"width\": 725, \"height\": 95, \"row\": {\"domain\": [\"+\", \"-\"], \"field\": \"strand\", \"type\": \"nominal\"}, \"color\": {\"domain\": [\"+\", \"-\"], \"field\": \"strand\", \"range\": [\"#7585FF\", \"#FF8A85\"], \"type\": \"nominal\"}, \"visibility\": [{\"operation\": \"less-than\", \"measure\": \"width\", \"threshold\": \"|xe-x|\", \"transitionPadding\": 10, \"target\": \"mark\"}], \"opacity\": {\"value\": 0.8}, \"id\": \"view1\"}, {\"alignment\": \"overlay\", \"tracks\": [{\"data\": {\"type\": \"beddb\", \"url\": \"http://localhost:19098/api/v1/tileset_info/?d=ab09f744-6932-4f43-a775-d5659509c531\", \"genomicFields\": [{\"index\": 1, \"name\": \"start\"}, {\"index\": 2, \"name\": \"end\"}], \"valueFields\": [{\"index\": 7, \"name\": \"significance\", \"type\": \"nominal\"}]}, \"mark\": \"bar\", \"color\": {\"value\": \"lightgray\"}, \"opacity\": {\"value\": 0.3}, \"size\": {\"value\": 1}, \"stroke\": {\"value\": \"lightgray\"}, \"strokeWidth\": {\"value\": 1}, \"visibility\": [{\"operation\": \"LT\", \"measure\": \"zoomLevel\", \"target\": \"mark\", \"threshold\": 100000, \"transitionPadding\": 100000}], \"x\": {\"field\": \"start\", \"type\": \"genomic\"}, \"y\": {\"baseline\": \"Uncertain_significance\", \"domain\": [\"Benign\", \"Benign/Likely_benign\", \"Likely_benign\", \"Uncertain_significance\", \"Likely_pathogenic\", \"Pathogenic/Likely_pathogenic\", \"Pathogenic\"], \"field\": \"significance\", \"legend\": true, \"range\": [214.28571428571428, 14.285714285714286], \"type\": \"nominal\"}, \"ye\": {\"value\": 100.0}}, {\"data\": {\"type\": \"beddb\", \"url\": \"http://localhost:19098/api/v1/tileset_info/?d=ab09f744-6932-4f43-a775-d5659509c531\", \"genomicFields\": [{\"index\": 1, \"name\": \"start\"}, {\"index\": 2, \"name\": \"end\"}], \"valueFields\": [{\"index\": 7, \"name\": \"significance\", \"type\": \"nominal\"}]}, \"mark\": \"point\", \"color\": {\"domain\": [\"Benign\", \"Benign/Likely_benign\", \"Likely_benign\", \"Uncertain_significance\", \"Likely_pathogenic\", \"Pathogenic/Likely_pathogenic\", \"Pathogenic\"], \"field\": \"significance\", \"range\": [\"#5A9F8C\", \"#5A9F8C\", \"#029F73\", \"gray\", \"#CB96B3\", \"#CB71A3\", \"#CB3B8C\"], \"type\": \"nominal\"}, \"opacity\": {\"value\": 0.8}, \"row\": {\"domain\": [\"Benign\", \"Benign/Likely_benign\", \"Likely_benign\", \"Uncertain_significance\", \"Likely_pathogenic\", \"Pathogenic/Likely_pathogenic\", \"Pathogenic\"], \"field\": \"significance\", \"type\": \"nominal\"}, \"size\": {\"value\": 7}, \"visibility\": [{\"operation\": \"LT\", \"measure\": \"zoomLevel\", \"target\": \"mark\", \"threshold\": 1000000, \"transitionPadding\": 1000000}], \"x\": {\"field\": \"start\", \"type\": \"genomic\"}}, {\"data\": {\"type\": \"multivec\", \"url\": \"http://localhost:19098/api/v1/tileset_info/?d=b5299802-ec63-47ab-9da0-7ffca093b2b1\", \"row\": \"significance\", \"column\": \"position\", \"value\": \"count\", \"categories\": [\"Benign\", \"Benign/Likely_benign\", \"Likely_benign\", \"Uncertain_significance\", \"Likely_pathogenic\", \"Pathogenic/Likely_pathogenic\", \"Pathogenic\"], \"binSize\": 4}, \"mark\": \"bar\", \"color\": {\"domain\": [\"Benign\", \"Benign/Likely_benign\", \"Likely_benign\", \"Uncertain_significance\", \"Likely_pathogenic\", \"Pathogenic/Likely_pathogenic\", \"Pathogenic\"], \"field\": \"significance\", \"legend\": true, \"range\": [\"#5A9F8C\", \"#5A9F8C\", \"#029F73\", \"gray\", \"#CB96B3\", \"#CB71A3\", \"#CB3B8C\"], \"type\": \"nominal\"}, \"visibility\": [{\"operation\": \"GT\", \"measure\": \"zoomLevel\", \"target\": \"mark\", \"threshold\": 500000, \"transitionPadding\": 500000}], \"x\": {\"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}, \"y\": {\"axis\": \"none\", \"field\": \"count\", \"type\": \"quantitative\"}}], \"width\": 150, \"height\": 200, \"xDomain\": {\"chromosome\": \"13\", \"interval\": [31500000, 33150000]}}], \"id\": \"view1\", \"xDomain\": {\"chromosome\": \"13\", \"interval\": [31500000, 33150000]}};\n",
       "    var opt = {\"padding\": 0, \"theme\": null};\n",
       "\n",
       "    loadGosling()\n",
       "        .then(gosling => gosling.embed(el, spec, opt))\n",
       "        .catch(err => {\n",
       "            el.innerHTML = `<div class=\"error\">\n",
       "    <p>JavaScript Error: ${error.message}</p>\n",
       "    <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>\n",
       "</div>`;\n",
       "            throw error;\n",
       "        });\n",
       "  </script>\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "View({\n",
       "  alignment: 'stack',\n",
       "  id: 'view1',\n",
       "  tracks: (View({\n",
       "    alignment: 'overlay',\n",
       "    color: Color({\n",
       "      domain: ['+', '-'],\n",
       "      field: 'strand',\n",
       "      range: ['#7585FF', '#FF8A85'],\n",
       "      type: 'nominal'\n",
       "    }),\n",
       "    height: 95,\n",
       "    id: 'view1',\n",
       "    opacity: {'value': 0.8},\n",
       "    row: Row({\n",
       "      domain: ['+', '-'],\n",
       "      field: 'strand',\n",
       "      type: 'nominal'\n",
       "    }),\n",
       "    tracks: (Track({\n",
       "      data: {'type': 'beddb', 'url': 'https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation', 'genomicFields': [{'index': 1, 'name': 'start'}, {'index': 2, 'name': 'end'}], 'valueFields': [{'index': 5, 'name': 'strand', 'type': 'nominal'}, {'index': 3, 'name': 'name', 'type': 'nominal'}], 'exonIntervalFields': [{'index': 12, 'name': 'start'}, {'index': 13, 'name': 'end'}]},\n",
       "      dataTransform: [FilterTransform({\n",
       "        field: 'type',\n",
       "        oneOf: ['gene'],\n",
       "        type: 'filter'\n",
       "      })],\n",
       "      mark: 'text',\n",
       "      style: Style({\n",
       "        dy: -15\n",
       "      }),\n",
       "      text: Text({\n",
       "        field: 'name',\n",
       "        type: 'nominal'\n",
       "      }),\n",
       "      x: X({\n",
       "        field: 'start',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      xe: Xe({\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      })\n",
       "    }), Track({\n",
       "      data: {'type': 'beddb', 'url': 'https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation', 'genomicFields': [{'index': 1, 'name': 'start'}, {'index': 2, 'name': 'end'}], 'valueFields': [{'index': 5, 'name': 'strand', 'type': 'nominal'}, {'index': 3, 'name': 'name', 'type': 'nominal'}], 'exonIntervalFields': [{'index': 12, 'name': 'start'}, {'index': 13, 'name': 'end'}]},\n",
       "      dataTransform: [FilterTransform({\n",
       "        field: 'type',\n",
       "        oneOf: ['gene'],\n",
       "        type: 'filter'\n",
       "      }), FilterTransform({\n",
       "        field: 'strand',\n",
       "        oneOf: ['+'],\n",
       "        type: 'filter'\n",
       "      })],\n",
       "      mark: 'triangleRight',\n",
       "      size: SizeValue({\n",
       "        value: 15\n",
       "      }),\n",
       "      x: X({\n",
       "        axis: 'top',\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      })\n",
       "    }), Track({\n",
       "      data: {'type': 'beddb', 'url': 'https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation', 'genomicFields': [{'index': 1, 'name': 'start'}, {'index': 2, 'name': 'end'}], 'valueFields': [{'index': 5, 'name': 'strand', 'type': 'nominal'}, {'index': 3, 'name': 'name', 'type': 'nominal'}], 'exonIntervalFields': [{'index': 12, 'name': 'start'}, {'index': 13, 'name': 'end'}]},\n",
       "      dataTransform: [FilterTransform({\n",
       "        field: 'type',\n",
       "        oneOf: ['gene'],\n",
       "        type: 'filter'\n",
       "      }), FilterTransform({\n",
       "        field: 'strand',\n",
       "        oneOf: ['-'],\n",
       "        type: 'filter'\n",
       "      })],\n",
       "      mark: 'triangleLeft',\n",
       "      size: SizeValue({\n",
       "        value: 15\n",
       "      }),\n",
       "      style: Style({\n",
       "        align: 'right'\n",
       "      }),\n",
       "      x: X({\n",
       "        field: 'start',\n",
       "        type: 'genomic'\n",
       "      })\n",
       "    }), Track({\n",
       "      data: {'type': 'beddb', 'url': 'https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation', 'genomicFields': [{'index': 1, 'name': 'start'}, {'index': 2, 'name': 'end'}], 'valueFields': [{'index': 5, 'name': 'strand', 'type': 'nominal'}, {'index': 3, 'name': 'name', 'type': 'nominal'}], 'exonIntervalFields': [{'index': 12, 'name': 'start'}, {'index': 13, 'name': 'end'}]},\n",
       "      dataTransform: [FilterTransform({\n",
       "        field: 'type',\n",
       "        oneOf: ['exon'],\n",
       "        type: 'filter'\n",
       "      })],\n",
       "      mark: 'rect',\n",
       "      size: SizeValue({\n",
       "        value: 15\n",
       "      }),\n",
       "      x: X({\n",
       "        field: 'start',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      xe: Xe({\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      })\n",
       "    }), Track({\n",
       "      data: {'type': 'beddb', 'url': 'https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation', 'genomicFields': [{'index': 1, 'name': 'start'}, {'index': 2, 'name': 'end'}], 'valueFields': [{'index': 5, 'name': 'strand', 'type': 'nominal'}, {'index': 3, 'name': 'name', 'type': 'nominal'}], 'exonIntervalFields': [{'index': 12, 'name': 'start'}, {'index': 13, 'name': 'end'}]},\n",
       "      dataTransform: [FilterTransform({\n",
       "        field: 'type',\n",
       "        oneOf: ['gene'],\n",
       "        type: 'filter'\n",
       "      }), FilterTransform({\n",
       "        field: 'strand',\n",
       "        oneOf: ['+'],\n",
       "        type: 'filter'\n",
       "      })],\n",
       "      mark: 'rule',\n",
       "      strokeWidth: StrokeWidthValue({\n",
       "        value: 3\n",
       "      }),\n",
       "      style: Style({\n",
       "        linePattern: {'type': 'triangleRight', 'size': 5}\n",
       "      }),\n",
       "      x: X({\n",
       "        field: 'start',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      xe: Xe({\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      })\n",
       "    }), Track({\n",
       "      data: {'type': 'beddb', 'url': 'https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation', 'genomicFields': [{'index': 1, 'name': 'start'}, {'index': 2, 'name': 'end'}], 'valueFields': [{'index': 5, 'name': 'strand', 'type': 'nominal'}, {'index': 3, 'name': 'name', 'type': 'nominal'}], 'exonIntervalFields': [{'index': 12, 'name': 'start'}, {'index': 13, 'name': 'end'}]},\n",
       "      dataTransform: [FilterTransform({\n",
       "        field: 'type',\n",
       "        oneOf: ['gene'],\n",
       "        type: 'filter'\n",
       "      }), FilterTransform({\n",
       "        field: 'strand',\n",
       "        oneOf: ['-'],\n",
       "        type: 'filter'\n",
       "      })],\n",
       "      mark: 'rule',\n",
       "      strokeWidth: StrokeWidthValue({\n",
       "        value: 3\n",
       "      }),\n",
       "      style: Style({\n",
       "        linePattern: {'type': 'triangleLeft', 'size': 5}\n",
       "      }),\n",
       "      x: X({\n",
       "        field: 'start',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      xe: Xe({\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      })\n",
       "    })),\n",
       "    visibility: [{'operation': 'less-than', 'measure': 'width', 'threshold': '|xe-x|', 'transitionPadding': 10, 'target': 'mark'}],\n",
       "    width: 725\n",
       "  }), View({\n",
       "    alignment: 'overlay',\n",
       "    height: 200,\n",
       "    tracks: (Track({\n",
       "      color: ColorValue({\n",
       "        value: 'lightgray'\n",
       "      }),\n",
       "      data: {'type': 'beddb', 'url': 'http://localhost:19098/api/v1/tileset_info/?d=ab09f744-6932-4f43-a775-d5659509c531', 'genomicFields': [{'index': 1, 'name': 'start'}, {'index': 2, 'name': 'end'}], 'valueFields': [{'index': 7, 'name': 'significance', 'type': 'nominal'}]},\n",
       "      mark: 'bar',\n",
       "      opacity: OpacityValue({\n",
       "        value: 0.3\n",
       "      }),\n",
       "      size: SizeValue({\n",
       "        value: 1\n",
       "      }),\n",
       "      stroke: StrokeValue({\n",
       "        value: 'lightgray'\n",
       "      }),\n",
       "      strokeWidth: StrokeWidthValue({\n",
       "        value: 1\n",
       "      }),\n",
       "      visibility: [VisibilityCondition({\n",
       "        measure: 'zoomLevel',\n",
       "        operation: 'LT',\n",
       "        target: 'mark',\n",
       "        threshold: 100000,\n",
       "        transitionPadding: 100000\n",
       "      })],\n",
       "      x: X({\n",
       "        field: 'start',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      y: Y({\n",
       "        baseline: 'Uncertain_significance',\n",
       "        domain: ['Benign', 'Benign/Likely_benign', 'Likely_benign', 'Uncertain_significance', 'Likely_pathogenic', 'Pathogenic/Likely_pathogenic', 'Pathogenic'],\n",
       "        field: 'significance',\n",
       "        legend: True,\n",
       "        range: [214.28571428571428, 14.285714285714286],\n",
       "        type: 'nominal'\n",
       "      }),\n",
       "      ye: YeValue({\n",
       "        value: 100.0\n",
       "      })\n",
       "    }), Track({\n",
       "      color: Color({\n",
       "        domain: ['Benign', 'Benign/Likely_benign', 'Likely_benign', 'Uncertain_significance', 'Likely_pathogenic', 'Pathogenic/Likely_pathogenic', 'Pathogenic'],\n",
       "        field: 'significance',\n",
       "        range: ['#5A9F8C', '#5A9F8C', '#029F73', 'gray', '#CB96B3', '#CB71A3', '#CB3B8C'],\n",
       "        type: 'nominal'\n",
       "      }),\n",
       "      data: {'type': 'beddb', 'url': 'http://localhost:19098/api/v1/tileset_info/?d=ab09f744-6932-4f43-a775-d5659509c531', 'genomicFields': [{'index': 1, 'name': 'start'}, {'index': 2, 'name': 'end'}], 'valueFields': [{'index': 7, 'name': 'significance', 'type': 'nominal'}]},\n",
       "      mark: 'point',\n",
       "      opacity: OpacityValue({\n",
       "        value: 0.8\n",
       "      }),\n",
       "      row: Row({\n",
       "        domain: ['Benign', 'Benign/Likely_benign', 'Likely_benign', 'Uncertain_significance', 'Likely_pathogenic', 'Pathogenic/Likely_pathogenic', 'Pathogenic'],\n",
       "        field: 'significance',\n",
       "        type: 'nominal'\n",
       "      }),\n",
       "      size: SizeValue({\n",
       "        value: 7\n",
       "      }),\n",
       "      visibility: [VisibilityCondition({\n",
       "        measure: 'zoomLevel',\n",
       "        operation: 'LT',\n",
       "        target: 'mark',\n",
       "        threshold: 1000000,\n",
       "        transitionPadding: 1000000\n",
       "      })],\n",
       "      x: X({\n",
       "        field: 'start',\n",
       "        type: 'genomic'\n",
       "      })\n",
       "    }), Track({\n",
       "      color: Color({\n",
       "        domain: ['Benign', 'Benign/Likely_benign', 'Likely_benign', 'Uncertain_significance', 'Likely_pathogenic', 'Pathogenic/Likely_pathogenic', 'Pathogenic'],\n",
       "        field: 'significance',\n",
       "        legend: True,\n",
       "        range: ['#5A9F8C', '#5A9F8C', '#029F73', 'gray', '#CB96B3', '#CB71A3', '#CB3B8C'],\n",
       "        type: 'nominal'\n",
       "      }),\n",
       "      data: {'type': 'multivec', 'url': 'http://localhost:19098/api/v1/tileset_info/?d=b5299802-ec63-47ab-9da0-7ffca093b2b1', 'row': 'significance', 'column': 'position', 'value': 'count', 'categories': ['Benign', 'Benign/Likely_benign', 'Likely_benign', 'Uncertain_significance', 'Likely_pathogenic', 'Pathogenic/Likely_pathogenic', 'Pathogenic'], 'binSize': 4},\n",
       "      mark: 'bar',\n",
       "      visibility: [VisibilityCondition({\n",
       "        measure: 'zoomLevel',\n",
       "        operation: 'GT',\n",
       "        target: 'mark',\n",
       "        threshold: 500000,\n",
       "        transitionPadding: 500000\n",
       "      })],\n",
       "      x: X({\n",
       "        field: 'start',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      xe: Xe({\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      y: Y({\n",
       "        axis: 'none',\n",
       "        field: 'count',\n",
       "        type: 'quantitative'\n",
       "      })\n",
       "    })),\n",
       "    width: 150,\n",
       "    xDomain: GenomicDomain({\n",
       "      chromosome: '13',\n",
       "      interval: [31500000, 33150000]\n",
       "    })\n",
       "  })),\n",
       "  xDomain: GenomicDomain({\n",
       "    chromosome: '13',\n",
       "    interval: [31500000, 33150000]\n",
       "  })\n",
       "})"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import gosling as gos\n",
    "import tracks\n",
    "\n",
    "lollipop_height = 200\n",
    "\n",
    "clin_var_beddb = gos.beddb(\n",
    "    \"../data/agg/clinvar.bed.beddb\",\n",
    "    genomicFields=[{\"index\": 1, \"name\": \"start\"}, {\"index\": 2, \"name\": \"end\"}],\n",
    "    valueFields=[{\"index\": 7, \"name\": \"significance\", \"type\": \"nominal\"}],\n",
    ")\n",
    "\n",
    "categories = [\n",
    "    \"Benign\", \"Benign/Likely_benign\", \"Likely_benign\", \n",
    "    \"Uncertain_significance\", \"Likely_pathogenic\", \n",
    "    \"Pathogenic/Likely_pathogenic\", \"Pathogenic\",\n",
    "]\n",
    "\n",
    "colors = [\"#5A9F8C\", \"#5A9F8C\", \"#029F73\", \"gray\", \"#CB96B3\", \"#CB71A3\", \"#CB3B8C\"]\n",
    "\n",
    "clin_var_multivec = gos.multivec(\n",
    "    \"../data/agg/density.multires.mv5\",\n",
    "    row=\"significance\",\n",
    "    column=\"position\",\n",
    "    value=\"count\",\n",
    "    categories=categories,\n",
    "    binSize=4,\n",
    ")\n",
    "\n",
    "dy = lollipop_height / len(categories) / 2 # workaround to offset\n",
    "strips = gos.Track(clin_var_beddb).mark_bar().encode(\n",
    "    x=\"start:G\",\n",
    "    y=gos.Y(\"significance:N\", domain=categories, range=[lollipop_height + dy, dy], baseline=\"Uncertain_significance\", legend=True),\n",
    "    ye=gos.value(lollipop_height/2),\n",
    "    size=gos.value(1),\n",
    "    color=gos.value(\"lightgray\"),\n",
    "    stroke=gos.value(\"lightgray\"),\n",
    "    strokeWidth=gos.value(1),\n",
    "    opacity=gos.value(0.3),\n",
    ").visibility_lt(\n",
    "    measure=\"zoomLevel\",\n",
    "    target=\"mark\",\n",
    "    threshold=100000,\n",
    "    transitionPadding=100000,\n",
    ")\n",
    "\n",
    "points = gos.Track(clin_var_beddb).mark_point().encode(\n",
    "    x=\"start:G\",\n",
    "    color=gos.Color(\"significance:N\", domain=categories, range=colors),\n",
    "    row=gos.Row(\"significance:N\", domain=categories),\n",
    "    size=gos.value(7),\n",
    "    opacity=gos.value(0.8),\n",
    ").visibility_lt(\n",
    "    measure=\"zoomLevel\",\n",
    "    target=\"mark\",\n",
    "    threshold=1000000,\n",
    "    transitionPadding=1000000,\n",
    ")\n",
    "\n",
    "bars = gos.Track(clin_var_multivec).mark_bar().encode(\n",
    "    x=\"start:G\",\n",
    "    xe=\"end:G\",\n",
    "    y=gos.Y(\"count:Q\", axis=\"none\"),\n",
    "    color=gos.Color(\"significance:N\", domain=categories, range=colors, legend=True)\n",
    ").visibility_gt(\n",
    "    measure=\"zoomLevel\",\n",
    "    target=\"mark\",\n",
    "    threshold=500000,\n",
    "    transitionPadding=500000,\n",
    ")\n",
    "\n",
    "lolipop = gos.overlay(strips, points, bars).properties(\n",
    "    width=725,\n",
    "    height=lollipop_height,\n",
    "    xDomain=gos.GenomicDomain(chromosome=\"13\", interval=[31500000, 33150000]),\n",
    ")\n",
    "\n",
    "# TODO: Need to specify top-level \n",
    "view = gos.stack(\n",
    "    tracks.gene_annotation.properties(id=\"view1\", height=95, width=725),\n",
    "    lolipop.properties(height=lollipop_height, width=150),\n",
    ").properties(\n",
    "    id=\"view1\",\n",
    "    xDomain=gos.GenomicDomain(chromosome=\"13\", interval=[31500000, 33150000]),\n",
    ")\n",
    "\n",
    "view"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dcdf3dc4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "621155c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8e5cb566b587411b9b9bacadf66ee5f2",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "GoslingWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "w = view.widget()\n",
    "w"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fcde7258",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
