{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2cbf8879",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<!DOCTYPE html>\n",
       "<html>\n",
       "<head>\n",
       "  <style>.error { color: red; }</style>\n",
       "  <link rel=\"stylesheet\" href=\"https://unpkg.com/higlass@1.11/dist/hglib.css\">\n",
       "</head>\n",
       "<body>\n",
       "  <div id=\"jupyter-gosling-f9f37b2dd4414ed9ac8bb07f9ca52273\"></div>\n",
       "  <script type=\"module\">\n",
       "\n",
       "    async function loadScript(src) {\n",
       "        return new Promise(resolve => {\n",
       "            const script = document.createElement('script');\n",
       "            script.onload = resolve;\n",
       "            script.src = src;\n",
       "            script.async = false;\n",
       "            document.head.appendChild(script);\n",
       "        });\n",
       "    }\n",
       "\n",
       "    async function loadGosling() {\n",
       "        // Manually load scripts from window namespace since requirejs might not be\n",
       "        // available in all browser environments.\n",
       "\n",
       "        // https://github.com/DanielHreben/requirejs-toggle\n",
       "        if (!window.gosling) {\n",
       "\n",
       "            // https://github.com/DanielHreben/requirejs-toggle\n",
       "            window.__requirejsToggleBackup = {\n",
       "                define: window.define,\n",
       "                require: window.require,\n",
       "                requirejs: window.requirejs,\n",
       "            };\n",
       "\n",
       "            for (const field of Object.keys(window.__requirejsToggleBackup)) {\n",
       "                window[field] = undefined;\n",
       "            }\n",
       "\n",
       "            // load dependencies sequentially\n",
       "            const sources = [\n",
       "                \"https://unpkg.com/react@17/umd/react.production.min.js\",\n",
       "                \"https://unpkg.com/react-dom@17/umd/react-dom.production.min.js\",\n",
       "                \"https://unpkg.com/pixi.js@6/dist/browser/pixi.min.js\",\n",
       "                \"https://unpkg.com/higlass@1.11/dist/hglib.js\",\n",
       "                \"https://unpkg.com/gosling.js@0.9.16/dist/gosling.js\",\n",
       "            ];\n",
       "\n",
       "            for (const src of sources) await loadScript(src);\n",
       "\n",
       "            // restore requirejs after scripts have loaded\n",
       "            Object.assign(window, window.__requirejsToggleBackup);\n",
       "            delete window.__requirejsToggleBackup;\n",
       "\n",
       "        }\n",
       "\n",
       "        return window.gosling;\n",
       "    };\n",
       "\n",
       "    var el = document.getElementById('jupyter-gosling-f9f37b2dd4414ed9ac8bb07f9ca52273');\n",
       "    var spec = {\"alignment\": \"overlay\", \"tracks\": [{\"data\": {\"type\": \"beddb\", \"url\": \"http://localhost:20978/api/v1/tileset_info/?d=00eed9cc-4d3b-4434-bb16-25d58a7e40a8\", \"genomicFields\": [{\"index\": 1, \"name\": \"start\"}, {\"index\": 2, \"name\": \"end\"}], \"valueFields\": [{\"index\": 7, \"name\": \"significance\", \"type\": \"nominal\"}]}, \"mark\": \"bar\", \"color\": {\"value\": \"lightgray\"}, \"opacity\": {\"value\": 0.3}, \"size\": {\"value\": 1}, \"stroke\": {\"value\": \"lightgray\"}, \"strokeWidth\": {\"value\": 1}, \"visibility\": [{\"operation\": \"LT\", \"measure\": \"zoomLevel\", \"target\": \"mark\", \"threshold\": 100000, \"transitionPadding\": 100000}], \"x\": {\"field\": \"start\", \"type\": \"genomic\"}, \"y\": {\"baseline\": \"Uncertain_significance\", \"domain\": [\"Benign\", \"Benign/Likely_benign\", \"Likely_benign\", \"Uncertain_significance\", \"Likely_pathogenic\", \"Pathogenic/Likely_pathogenic\", \"Pathogenic\"], \"field\": \"significance\", \"range\": [150, 20], \"type\": \"nominal\"}}, {\"data\": {\"type\": \"beddb\", \"url\": \"http://localhost:20978/api/v1/tileset_info/?d=00eed9cc-4d3b-4434-bb16-25d58a7e40a8\", \"genomicFields\": [{\"index\": 1, \"name\": \"start\"}, {\"index\": 2, \"name\": \"end\"}], \"valueFields\": [{\"index\": 7, \"name\": \"significance\", \"type\": \"nominal\"}]}, \"mark\": \"point\", \"color\": {\"domain\": [\"Benign\", \"Benign/Likely_benign\", \"Likely_benign\", \"Uncertain_significance\", \"Likely_pathogenic\", \"Pathogenic/Likely_pathogenic\", \"Pathogenic\"], \"field\": \"significance\", \"range\": [\"#5A9F8C\", \"#5A9F8C\", \"#029F73\", \"gray\", \"#CB96B3\", \"#CB71A3\", \"#CB3B8C\"], \"type\": \"nominal\"}, \"opacity\": {\"value\": 0.8}, \"row\": {\"domain\": [\"Benign\", \"Benign/Likely_benign\", \"Likely_benign\", \"Uncertain_significance\", \"Likely_pathogenic\", \"Pathogenic/Likely_pathogenic\", \"Pathogenic\"], \"field\": \"significance\", \"type\": \"nominal\"}, \"size\": {\"value\": 7}, \"visibility\": [{\"operation\": \"LT\", \"measure\": \"zoomLevel\", \"target\": \"mark\", \"threshold\": 1000000, \"transitionPadding\": 1000000}], \"x\": {\"field\": \"start\", \"type\": \"genomic\"}}, {\"data\": {\"type\": \"multivec\", \"url\": \"http://localhost:20978/api/v1/tileset_info/?d=6aac6c95-bb6d-4728-b5c4-cd8ce17e698c\", \"row\": \"significance\", \"column\": \"position\", \"value\": \"count\", \"categories\": [\"Benign\", \"Benign/Likely_benign\", \"Likely_benign\", \"Uncertain_significance\", \"Likely_pathogenic\", \"Pathogenic/Likely_pathogenic\", \"Pathogenic\"], \"binSize\": 4}, \"mark\": \"bar\", \"color\": {\"domain\": [\"Benign\", \"Benign/Likely_benign\", \"Likely_benign\", \"Uncertain_significance\", \"Likely_pathogenic\", \"Pathogenic/Likely_pathogenic\", \"Pathogenic\"], \"field\": \"significance\", \"legend\": true, \"range\": [\"#5A9F8C\", \"#5A9F8C\", \"#029F73\", \"gray\", \"#CB96B3\", \"#CB71A3\", \"#CB3B8C\"], \"type\": \"nominal\"}, \"visibility\": [{\"operation\": \"GT\", \"measure\": \"zoomLevel\", \"target\": \"mark\", \"threshold\": 500000, \"transitionPadding\": 500000}], \"x\": {\"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}, \"y\": {\"axis\": \"none\", \"field\": \"count\", \"type\": \"quantitative\"}}], \"width\": 725, \"height\": 150, \"xDomain\": {\"chromosome\": \"13\", \"interval\": [31500000, 33150000]}};\n",
       "    var opt = {\"padding\": 0, \"theme\": null};\n",
       "\n",
       "    loadGosling()\n",
       "        .then(gosling => gosling.embed(el, spec, opt))\n",
       "        .catch(err => {\n",
       "            el.innerHTML = `<div class=\"error\">\n",
       "    <p>JavaScript Error: ${error.message}</p>\n",
       "    <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>\n",
       "</div>`;\n",
       "            throw error;\n",
       "        });\n",
       "  </script>\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "View({\n",
       "  alignment: 'overlay',\n",
       "  height: 150,\n",
       "  tracks: (Track({\n",
       "    color: ColorValue({\n",
       "      value: 'lightgray'\n",
       "    }),\n",
       "    data: {'type': 'beddb', 'url': 'http://localhost:20978/api/v1/tileset_info/?d=00eed9cc-4d3b-4434-bb16-25d58a7e40a8', 'genomicFields': [{'index': 1, 'name': 'start'}, {'index': 2, 'name': 'end'}], 'valueFields': [{'index': 7, 'name': 'significance', 'type': 'nominal'}]},\n",
       "    mark: 'bar',\n",
       "    opacity: OpacityValue({\n",
       "      value: 0.3\n",
       "    }),\n",
       "    size: SizeValue({\n",
       "      value: 1\n",
       "    }),\n",
       "    stroke: StrokeValue({\n",
       "      value: 'lightgray'\n",
       "    }),\n",
       "    strokeWidth: StrokeWidthValue({\n",
       "      value: 1\n",
       "    }),\n",
       "    visibility: [VisibilityCondition({\n",
       "      measure: 'zoomLevel',\n",
       "      operation: 'LT',\n",
       "      target: 'mark',\n",
       "      threshold: 100000,\n",
       "      transitionPadding: 100000\n",
       "    })],\n",
       "    x: X({\n",
       "      field: 'start',\n",
       "      type: 'genomic'\n",
       "    }),\n",
       "    y: Y({\n",
       "      baseline: 'Uncertain_significance',\n",
       "      domain: ['Benign', 'Benign/Likely_benign', 'Likely_benign', 'Uncertain_significance', 'Likely_pathogenic', 'Pathogenic/Likely_pathogenic', 'Pathogenic'],\n",
       "      field: 'significance',\n",
       "      range: [150, 20],\n",
       "      type: 'nominal'\n",
       "    })\n",
       "  }), Track({\n",
       "    color: Color({\n",
       "      domain: ['Benign', 'Benign/Likely_benign', 'Likely_benign', 'Uncertain_significance', 'Likely_pathogenic', 'Pathogenic/Likely_pathogenic', 'Pathogenic'],\n",
       "      field: 'significance',\n",
       "      range: ['#5A9F8C', '#5A9F8C', '#029F73', 'gray', '#CB96B3', '#CB71A3', '#CB3B8C'],\n",
       "      type: 'nominal'\n",
       "    }),\n",
       "    data: {'type': 'beddb', 'url': 'http://localhost:20978/api/v1/tileset_info/?d=00eed9cc-4d3b-4434-bb16-25d58a7e40a8', 'genomicFields': [{'index': 1, 'name': 'start'}, {'index': 2, 'name': 'end'}], 'valueFields': [{'index': 7, 'name': 'significance', 'type': 'nominal'}]},\n",
       "    mark: 'point',\n",
       "    opacity: OpacityValue({\n",
       "      value: 0.8\n",
       "    }),\n",
       "    row: Row({\n",
       "      domain: ['Benign', 'Benign/Likely_benign', 'Likely_benign', 'Uncertain_significance', 'Likely_pathogenic', 'Pathogenic/Likely_pathogenic', 'Pathogenic'],\n",
       "      field: 'significance',\n",
       "      type: 'nominal'\n",
       "    }),\n",
       "    size: SizeValue({\n",
       "      value: 7\n",
       "    }),\n",
       "    visibility: [VisibilityCondition({\n",
       "      measure: 'zoomLevel',\n",
       "      operation: 'LT',\n",
       "      target: 'mark',\n",
       "      threshold: 1000000,\n",
       "      transitionPadding: 1000000\n",
       "    })],\n",
       "    x: X({\n",
       "      field: 'start',\n",
       "      type: 'genomic'\n",
       "    })\n",
       "  }), Track({\n",
       "    color: Color({\n",
       "      domain: ['Benign', 'Benign/Likely_benign', 'Likely_benign', 'Uncertain_significance', 'Likely_pathogenic', 'Pathogenic/Likely_pathogenic', 'Pathogenic'],\n",
       "      field: 'significance',\n",
       "      legend: True,\n",
       "      range: ['#5A9F8C', '#5A9F8C', '#029F73', 'gray', '#CB96B3', '#CB71A3', '#CB3B8C'],\n",
       "      type: 'nominal'\n",
       "    }),\n",
       "    data: {'type': 'multivec', 'url': 'http://localhost:20978/api/v1/tileset_info/?d=6aac6c95-bb6d-4728-b5c4-cd8ce17e698c', 'row': 'significance', 'column': 'position', 'value': 'count', 'categories': ['Benign', 'Benign/Likely_benign', 'Likely_benign', 'Uncertain_significance', 'Likely_pathogenic', 'Pathogenic/Likely_pathogenic', 'Pathogenic'], 'binSize': 4},\n",
       "    mark: 'bar',\n",
       "    visibility: [VisibilityCondition({\n",
       "      measure: 'zoomLevel',\n",
       "      operation: 'GT',\n",
       "      target: 'mark',\n",
       "      threshold: 500000,\n",
       "      transitionPadding: 500000\n",
       "    })],\n",
       "    x: X({\n",
       "      field: 'start',\n",
       "      type: 'genomic'\n",
       "    }),\n",
       "    xe: Xe({\n",
       "      field: 'end',\n",
       "      type: 'genomic'\n",
       "    }),\n",
       "    y: Y({\n",
       "      axis: 'none',\n",
       "      field: 'count',\n",
       "      type: 'quantitative'\n",
       "    })\n",
       "  })),\n",
       "  width: 725,\n",
       "  xDomain: GenomicDomain({\n",
       "    chromosome: '13',\n",
       "    interval: [31500000, 33150000]\n",
       "  })\n",
       "})"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import gosling as gos\n",
    " \n",
    "clin_var_beddb = gos.beddb(\n",
    "    \"../data/agg/clinvar.bed.beddb\",\n",
    "    genomicFields=[{\"index\": 1, \"name\": \"start\"}, {\"index\": 2, \"name\": \"end\"}],\n",
    "    valueFields=[{\"index\": 7, \"name\": \"significance\", \"type\": \"nominal\"}],\n",
    ")\n",
    "\n",
    "categories = [\n",
    "    \"Benign\", \"Benign/Likely_benign\", \"Likely_benign\", \n",
    "    \"Uncertain_significance\", \"Likely_pathogenic\", \n",
    "    \"Pathogenic/Likely_pathogenic\", \"Pathogenic\",\n",
    "]\n",
    "\n",
    "colors = [\"#5A9F8C\", \"#5A9F8C\", \"#029F73\", \"gray\", \"#CB96B3\", \"#CB71A3\", \"#CB3B8C\"]\n",
    "\n",
    "clin_var_multivec = gos.multivec(\n",
    "    \"../data/agg/density.multires.mv5\",\n",
    "    row=\"significance\",\n",
    "    column=\"position\",\n",
    "    value=\"count\",\n",
    "    categories=categories,\n",
    "    binSize=4,\n",
    ")\n",
    "\n",
    "\n",
    "lolipop = gos.overlay(\n",
    "    (\n",
    "        gos.Track(clin_var_beddb).mark_bar().encode(\n",
    "            x=\"start:G\",\n",
    "            y=gos.Y(\"significance:N\", domain=categories, range=[150, 20], baseline=\"Uncertain_significance\"),\n",
    "            size=gos.value(1),\n",
    "            color=gos.value(\"lightgray\"),\n",
    "            stroke=gos.value(\"lightgray\"),\n",
    "            strokeWidth=gos.value(1),\n",
    "            opacity=gos.value(0.3),\n",
    "        ).visibility_lt(\n",
    "            measure=\"zoomLevel\",\n",
    "            target=\"mark\",\n",
    "            threshold=100000,\n",
    "            transitionPadding=100000,\n",
    "        )\n",
    "    ),\n",
    "    (\n",
    "        gos.Track(clin_var_beddb).mark_point().encode(\n",
    "            x=\"start:G\",\n",
    "            color=gos.Color(\"significance:N\", domain=categories, range=colors),\n",
    "            row=gos.Row(\"significance:N\", domain=categories),\n",
    "            size=gos.value(7),\n",
    "            opacity=gos.value(0.8),\n",
    "        ).visibility_lt(\n",
    "            measure=\"zoomLevel\",\n",
    "            target=\"mark\",\n",
    "            threshold=1000000,\n",
    "            transitionPadding=1000000,\n",
    "        )\n",
    "    ),\n",
    "    (\n",
    "        gos.Track(clin_var_multivec).mark_bar().encode(\n",
    "            x=\"start:G\",\n",
    "            xe=\"end:G\",\n",
    "            y=gos.Y(\"count:Q\", axis=\"none\"),\n",
    "            color=gos.Color(\"significance:N\", domain=categories, range=colors, legend=True)\n",
    "        ).visibility_gt(\n",
    "            measure=\"zoomLevel\",\n",
    "            target=\"mark\",\n",
    "            threshold=500000,\n",
    "            transitionPadding=500000,\n",
    "        )\n",
    "    ),\n",
    "    width=725,\n",
    "    height=150,\n",
    "    xDomain=gos.GenomicDomain(chromosome=\"13\", interval=[31500000, 33150000]),\n",
    ")\n",
    "\n",
    "lolipop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b76fb149",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
